### PR TITLE
Update "dtype" to "type" in ParamSpec dictionary representation

### PR DESCRIPTION
**Pull request by serena-ruan**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?
Previously ParamSpec's to_dict method uses "dtype" for field ParamSpec.dtype, to be consistent with ColSpec and TensorSpec, we switched to use "type" in the dictionary representation.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/types/schema.py [hunk 1]
```diff
@@ -685,16 +685,23 @@ def __repr__(self) -> str:
...
-        The dictionary is expected to contain `name`, `dtype` and `default` keys.
+        The dictionary is expected to contain `name`, `type` and `default` keys.
         """
-        if not {"name", "dtype", "default"} <= set(kwargs.keys()):
+        # For backward compatibility, we accept both `type` and `dtype` keys
+        if not {"name", "dtype", "default"} <= set(kwargs.keys()) and not {
+            "name",
+            "type",
+            "default",
+        } <= set(kwargs.keys()):
```
Comment (harupy) [inline mlflow/types/schema.py:695]: /ai is it possible to simplify this condition?

Comment (harupy) [reply mlflow/types/schema.py:695]: @serena-ruan Can we make this change?

Comment (serena-ruan) [reply mlflow/types/schema.py:695]: @harupy Do you have other comments? I think this https://github.com/mlflow/mlflow/pull/9565#discussion_r1319556184 seems even more complicated

Comment (WeichenXu123) [reply mlflow/types/schema.py:695]: I prefers calling `issubet` which is more clear. :)

File: tests/types/test_schema.py [hunk 2]
```diff
@@ -876,13 +876,26 @@ def test_param_spec_to_and_from_dict():
...
         "name": "datetime_array",
-        "dtype": "datetime",
+        "type": "datetime",
         "default": ["2023-06-26T00:00:00", "2023-06-27T00:00:00"],
         "shape": (-1,),
     }
     assert ParamSpec.from_json_dict(**json.loads(json.dumps(spec.to_dict()))) == spec
 
 
+def test_param_spec_from_dict_back_compat():
```
Comment (WeichenXu123) [inline tests/types/test_schema.py:886]: nit:
```suggestion
def test_param_spec_from_dict_backward_compatibility():
```

**Review by WeichenXu123 — APPROVED** [review]

**Pull request merged**